### PR TITLE
[GR-34108] Extend boot module layer to include required modules

### PR DESCRIPTION
**Pull request by ivan-ristovic**
This PR implements #3733

### CODE REVIEW

File: substratevm/src/com.oracle.svm.hosted.jdk11/src/com/oracle/svm/hosted/ModuleLayerFeature.java [hunk 1]
```diff
@@ -119,49 +139,183 @@ public void afterAnalysis(AfterAnalysisAccess access) {
...
+            }
+        }
+
+        return nameToModule;
+    }
+
+    /*
+     * Creates a stream of module names that are reachable from a given module through "requires"
+     */
+    private static Stream<String> extractRequiredModuleNames(Module m) {
```
Comment (olpaw) [inline substratevm/src/com.oracle.svm.hosted.jdk11/src/com/oracle/svm/hosted/ModuleLayerFeature.java:316]: Maybe move next to caller (right after `afterAnalysis`)

File: substratevm/src/com.oracle.svm.hosted.jdk11/src/com/oracle/svm/hosted/ModuleLayerFeature.java [hunk 2]
```diff
@@ -119,49 +139,183 @@ public void afterAnalysis(AfterAnalysisAccess access) {
...
         }
+
+        // Ensure that the lazy modules field gets set
+        runtimeBootLayer.modules();
+    }
+
+    /**
+     * This method creates Module instances that will populate the runtime boot module layer of the image.
+     * This implementation is copy-pasted from {@link java.lang.Module#defineModules(Configuration, Function, ModuleLayer)}
+     * with few simplifications (removing multiple classloader support) and removal of VM state updates
```
Comment (olpaw) [inline substratevm/src/com.oracle.svm.hosted.jdk11/src/com/oracle/svm/hosted/ModuleLayerFeature.java:195]: > removal of VM state updates

:grimacing: if `java.lang.Module#defineModules` would have been written in a way that partitions module creation from VM state update we could reuse the module creation part without this reflection-orgy.

**Review by olpaw — CHANGES_REQUESTED** [review]

File: substratevm/src/com.oracle.svm.hosted.jdk11/src/com/oracle/svm/hosted/ModuleLayerFeature.java [hunk 3]
```diff
@@ -110,7 +130,7 @@ public void beforeAnalysis(BeforeAnalysisAccess access) {
         Map<String, Module> baseModules = ModuleLayer.boot().modules()
                 .stream()
                 .collect(Collectors.toMap(Module::getName, m -> m));
```
[the remark this answers is not in the document]
Comment (ivan-ristovic) [reply substratevm/src/com.oracle.svm.hosted.jdk11/src/com/oracle/svm/hosted/ModuleLayerFeature.java]: Fixed.

File: substratevm/src/com.oracle.svm.hosted.jdk11/src/com/oracle/svm/hosted/ModuleLayerFeature.java (hunk 1 above)
Comment (ivan-ristovic) [reply substratevm/src/com.oracle.svm.hosted.jdk11/src/com/oracle/svm/hosted/ModuleLayerFeature.java]: Done.

File: substratevm/src/com.oracle.svm.hosted.jdk11/src/com/oracle/svm/hosted/ModuleLayerFeature.java [hunk 4]
```diff
@@ -99,9 +111,17 @@ public boolean isInConfiguration(IsInConfigurationAccess access) {
     @Override
     public void afterRegistration(AfterRegistrationAccess access) {
         ImageSingletons.add(BootModuleLayerSupport.class, new BootModuleLayerSupport());
-        moduleNameToModuleField = ReflectionUtil.lookupField(ModuleLayer.class, "nameToModule");
-        moduleParentsField = ReflectionUtil.lookupField(ModuleLayer.class, "parents");
+        everyoneModule = ReflectionUtil.readField(Module.class, "EVERYONE_MODULE", null);
```
[the remark this answers is not in the document]
Comment (ivan-ristovic) [reply substratevm/src/com.oracle.svm.hosted.jdk11/src/com/oracle/svm/hosted/ModuleLayerFeature.java]: Done.

**Review by olpaw — CHANGES_REQUESTED** [review]
Looks good now.

Please also run `mx eclipseformat --primary` to see where you need to adjust the formatting.

**Comment by olpaw** [comment]
Unfortunately it e.g. fails `mx gate --omit-clean -t 'native unittests'`
...
```
[com.oracle.svm.junit.svmjunitrunner:5896]    classlist:   1,336.36 ms,  0.96 GB
[com.oracle.svm.junit.svmjunitrunner:5896]        (cap):     744.56 ms,  0.96 GB
[com.oracle.svm.junit.svmjunitrunner:5896]        setup:   4,608.89 ms,  0.96 GB
[com.oracle.svm.junit.svmjunitrunner:5896]     (clinit):     464.64 ms,  3.22 GB
[com.oracle.svm.junit.svmjunitrunner:5896]     analysis:  23,231.12 ms,  3.22 GB
Fatal error: com.oracle.svm.core.util.VMError$HostedError: Failed to synthesize the runtime boot module layer configuration.
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.util.VMError.shouldNotReachHere(VMError.java:76)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.ModuleLayerFeature.synthesizeRuntimeBootLayerConfiguration(ModuleLayerFeature.java:170)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.ModuleLayerFeature.synthesizeRuntimeBootLayer(ModuleLayerFeature.java:153)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.ModuleLayerFeature.afterAnalysis(ModuleLayerFeature.java:140)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.lambda$runPointsToAnalysis$13(NativeImageGenerator.java:752)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.FeatureHandler.forEachFeature(FeatureHandler.java:73)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:752)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:532)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:491)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:415)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:573)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:126)
Caused by: java.lang.module.FindException: Module junit not found
	at java.base/java.lang.module.Resolver.findFail(Resolver.java:877)
	at java.base/java.lang.module.Resolver.resolve(Resolver.java:128)
	at java.base/java.lang.module.Configuration.resolve(Configuration.java:422)
	at java.base/java.lang.module.Configuration.resolve(Configuration.java:256)
[com.oracle.svm.junit.svmjunitrunner:5896]      [total]:  29,619.83 ms,  3.22 GB
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.ModuleLayerFeature.synthesizeRuntimeBootLayerConfiguration(ModuleLayerFeature.java:168)
	... 10 more
```
The current solution is not able to deal with optional required modules (`junit` in this case).
See https://github.com/oracle/graal/blob/f49b9d19b1496b701d2b60053311553605d6349a/substratevm/mx.substratevm/suite.py#L1237

Those are optional dependencies. I.e. we have to be able to handle the situation that these modules are not necessarily available during image build time.

**Comment by ivan-ristovic** [comment]
Style issues fixed.

I believe the reflection error that you sent is something we have seen earlier and is related to it behaving differently on 16+. If you recall we discussed a similar "reflection bug" where the field is present in a class and reflection for some reason does not pick it up. Is this the same issue then, because as far as I see there is no reason why `EVERYONE_MODULE` cannot be accessed via reflection (it is present in the `Module` class in the same way as on 11)?

**Comment by olpaw** [comment]
@ivan-ristovic CI passes now mostly. Except for one little style issue:
```
1. ERROR in /b/b/e/main/substratevm/src/com.oracle.svm.hosted.jdk11/src/com/oracle/svm/hosted/ModuleLayerFeature.java (at line 258)
	private Field findFieldByName(Field[] fields, String name) {
	              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
The method findFieldByName(Field[], String) from the type ModuleLayerFeature.NameToModuleSynthesizer can be declared as static
----------
1 problem (1 error)
```

**Pull request merged**